### PR TITLE
Improve arrow flight batch splitting and naming

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

re #3371 

# Rationale for this change
 
I wrote some tests for this code in  https://github.com/influxdata/influxdb_iox/pull/6484 (actually covers a bug I fixed when upstreaming to arrow-rs)

Also, it would be nice to clarify that the units are in bytes and not some other unit (like rows, for example) as suggested by @carols10cents on https://github.com/influxdata/influxdb_iox/pull/6460

# What changes are included in this PR?

1. Add more tests for record batch splitting
2. Change name of parameter from `max_message_size` to `max_message_size_bytes`

# Are there any user-facing changes?

No : Note that this code has not been released yet, so there are no external API changes.

### CODE REVIEW

File: arrow-flight/src/encode.rs [hunk 1]
```diff
@@ -100,8 +100,8 @@ impl FlightDataEncoderBuilder {
     /// is approximate because there additional encoding overhead on
     /// top of the underlying data itself.
     ///
-    pub fn with_max_message_size(mut self, max_batch_size: usize) -> Self {
-        self.max_batch_size = max_batch_size;
+    pub fn with_max_message_size_bytes(mut self, max_batch_size_bytes: usize) -> Self {
+        self.max_batch_size_bytes = max_batch_size_bytes;
```
Comment (tustvold) [inline arrow-flight/src/encode.rs:104]: I'm confused is this max message size of max batch size?

IMO it isn't entirely clear what a message means in this context, perhaps `max_flight_data_size` might be more clear?

Comment (alamb) [reply arrow-flight/src/encode.rs:104]: I like that -- will change

File: arrow-flight/src/encode.rs [hunk 2]
```diff
@@ -63,24 +63,25 @@ use futures::{ready, stream::BoxStream, Stream, StreamExt};
 /// [`FlightError`]: crate::error::FlightError
 #[derive(Debug)]
 pub struct FlightDataEncoderBuilder {
-    /// The maximum message size (see details on [`Self::with_max_message_size`]).
-    max_batch_size: usize,
+    /// The maximum approximate target message size in bytes
+    /// (see details on [`Self::with_max_flight_data_size`]).
+    max_flight_data_size: usize,
```
Comment (alamb) [inline arrow-flight/src/encode.rs:68]: I renamed this to `max_flight_data_size` and tried to clarify the intent more

**Review by tustvold — APPROVED** [review]

File: arrow-flight/src/encode.rs [hunk 3]
```diff
@@ -506,6 +512,48 @@ mod tests {
...
+
+    /// Creates a UInt64Array of 8 byte integers with input_rows rows
+    /// `max_flight_data_size_bytes` pieces and verifies the row counts in
+    /// those pieces
+    fn verify_split(
+        num_input_rows: u64,
+        max_flight_data_size_bytes: usize,
+        expected_sizes: Vec<usize>,
+    ) {
+        let array: UInt64Array = (0..num_input_rows).collect();
```
Comment (tustvold) [inline arrow-flight/src/encode.rs:541]: It occurs to me that this splitting won't work for dictionaries, not really sure what we can do about that though... :thinking:

Comment (alamb) [reply arrow-flight/src/encode.rs:541]: 🤔  I guess we hope for small dictionaries

**Pull request merged**